src/Adapter/Product/CommandHandler/GenerateProductCombinationsHandler.php

```diff
@@ -0,0 +1,185 @@
+        $this->dbInstance = Db::getInstance();
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @todo: multistore
+     */
+    public function handle(GenerateProductCombinationsCommand $command): array
+    {
+        $product = $this->getProduct($command->getProductId());
+        $generatedCombinations = $this->combinationGenerator->generate($command->getGroupedAttributeIds());
+
+        SpecificPriceRule::disableAnyApplication();
+
+        $combinationIds = $this->addCombinations($product, $generatedCombinations);
+
+        Product::updateDefaultAttribute($product->id);
+        SpecificPriceRule::enableAnyApplication();
+        SpecificPriceRule::applyAllRules([$product->id]);
+
+        return $combinationIds;
+    }
+
+    /**
+     * @param Product $product
+     * @param Iterator $generatedCombinations
+     *
+     * @return CombinationId[]
+     *
+     * @throws CombinationException
+     */
+    private function addCombinations(Product $product, Iterator $generatedCombinations): array
+    {
+        $addedCombinationIds = [];
+        foreach ($generatedCombinations as $generatedCombination) {
+            $combinationExists = $product->productAttributeExists($generatedCombination, false, null, true);
+
+            if ($combinationExists) {
+                continue;
+            }
+
+            try {
+                $addedCombinationIds[] = $this->addCombination($product, $generatedCombination);
+            } catch (PrestaShopException $e) {
+                throw new CombinationException(
+                    sprintf(
+                        'Error occurred when trying to add combination for product #%d. Combination: %s',
+                        $product->id,
+                        var_export($generatedCombination, true)
+                    ),
+                    $e
+                );
+            }
+        }
+
+        return $addedCombinationIds;
+    }
+
+    /**
+     * @param Product $product
+     * @param int[] $generatedCombination
+     *
+     * @return CombinationId
+     */
+    private function addCombination(Product $product, array $generatedCombination): CombinationId
+    {
+        $newCombination = new Combination();
+        $newCombination->id_product = $product->id;
+        $newCombination->default_on = 0;
+
+        $this->dbInstance->beginTransaction();
+        $product->setAvailableDate();
+
+        if (!$newCombination->add()) {
+            throw new CannotAddCombinationException(sprintf(
+                'Failed adding new combination for product #%d. Combination: %s',
+                $product->id,
+                var_export($generatedCombination, true)
+            ));
+        }
+        $combinationId = (int) $newCombination->id;
+
+        try {
+            $this->saveProductAttributeAssociation($combinationId, $generatedCombination);
+        } catch (Exception $e) {
+            $this->dbInstance->rollback();
```

Comment:
This try/catch for rollback should englobe the whole transaction (start right before beginTransaction, and and right after commit) So that any exception will trigger the rollback

Reply:
I've put $product->setAvailableDate outside the loop, so there is no more sql actions inside, meaining I dont need to wrap a transaction anywhere further